From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: fgei <fgei@gmail.com>
Date: Wed, 26 Feb 2025 09:44:24 +0000
Subject: [PATCH] Support for using constant webview provider class name

The prefix "WebViewChromiumFactoryProviderFor" is necessary to avoid the
need to add more proguard rules. This changeset adds support for
including an implementation of WebViewChromiumFactoryProvider with a
non-changing class name for usage in current and future AOSP-based OS,
and currently enabled by default. Once the corresponding OS change
enables the flag for using the declared classname in this patch, it will
always initialize the declared class name here regardless of upstream
config flag.
---
 android_webview/glue/BUILD.gn                 |  4 +++
 ...iewChromiumFactoryProviderForVanadium.java | 14 ++++++++++
 android_webview/glue/webview_glue_ext.gni     | 28 +++++++++++++++++++
 3 files changed, 46 insertions(+)
 create mode 100644 android_webview/glue/java/src/com/android/webview/chromium/WebViewChromiumFactoryProviderForVanadium.java
 create mode 100644 android_webview/glue/webview_glue_ext.gni

diff --git a/android_webview/glue/BUILD.gn b/android_webview/glue/BUILD.gn
index 81fccd1aba106..177b4ae1fe927 100644
--- a/android_webview/glue/BUILD.gn
+++ b/android_webview/glue/BUILD.gn
@@ -95,4 +95,8 @@ android_library("glue_java") {
     "java/src/com/android/webview/chromium/WebViewRenderProcessClientAdapter.java",
     "java/src/com/android/webview/chromium/WebkitToSharedGlueConverter.java",
   ]
+
+  import("//android_webview/glue/webview_glue_ext.gni")
+  sources += webview_glue_ext_java_sources
+  deps += webview_glue_ext_java_deps
 }
diff --git a/android_webview/glue/java/src/com/android/webview/chromium/WebViewChromiumFactoryProviderForVanadium.java b/android_webview/glue/java/src/com/android/webview/chromium/WebViewChromiumFactoryProviderForVanadium.java
new file mode 100644
index 0000000000000..4730d053a8492
--- /dev/null
+++ b/android_webview/glue/java/src/com/android/webview/chromium/WebViewChromiumFactoryProviderForVanadium.java
@@ -0,0 +1,14 @@
+package com.android.webview.chromium;
+
+import org.chromium.android_webview.common.Lifetime;
+
+@Lifetime.Singleton
+class WebViewChromiumFactoryProviderForVanadium extends WebViewChromiumFactoryProvider {
+    public static WebViewChromiumFactoryProvider create(android.webkit.WebViewDelegate delegate) {
+        return new WebViewChromiumFactoryProviderForVanadium(delegate);
+    }
+
+    protected WebViewChromiumFactoryProviderForVanadium(android.webkit.WebViewDelegate delegate) {
+        super(delegate);
+    }
+}
diff --git a/android_webview/glue/webview_glue_ext.gni b/android_webview/glue/webview_glue_ext.gni
new file mode 100644
index 0000000000000..6f346921723d7
--- /dev/null
+++ b/android_webview/glue/webview_glue_ext.gni
@@ -0,0 +1,28 @@
+# Copyright 2025 GrapheneOS
+# Use of this source code is governed by a GPL 2.0-style license that can be
+# found in the LICENSE file.
+
+import("//build/config/android/config.gni")
+import("//build/config/android/rules.gni")
+
+declare_args() {
+  use_constant_webview_provider_name = true
+}
+
+webview_glue_ext_java_sources = [
+]
+
+webview_glue_ext_java_resources = [
+]
+
+webview_glue_ext_java_deps = [
+]
+
+webview_glue_ext_java_srcjar_deps = [
+]
+
+if (use_constant_webview_provider_name) {
+  webview_glue_ext_java_sources += [
+    "java/src/com/android/webview/chromium/WebViewChromiumFactoryProviderForVanadium.java",
+  ]
+}
